docs/ksm.md

```diff
@@ -0,0 +1,129 @@
+# Kernel Same-Page ([KSM](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/vm/ksm.txt)) tuning
+
+This section describes why `cc-proxy` tunes the host kernel
+KSM settings and how it does it.
+
+## What is KSM ?
+
+KSM is a host Linux kernel feature for de-duplicating memory pages.
+Although it was initially designed as a KVM specific feature, it is
+now part of the generic Linux memory management subsystem and can
+be leveraged by any userspace component or application looking for
+memory to save.
+
+A daemon (ksmd) periodically scans userspace memory, looking for
+identical pages that can be replaced by a single, write-protected
+page. When a process tries to modify this shared page content, it
+gets a private copy into its memory space. KSM only scans and merges
+pages that are both anonymous and that have been explictly tagged as
+mergeable by applications calling into the madvise system call
+(`int madvice(addr, length, MADV_MERGEABLE)`).
+
+KSM is tunable through a set of Linux kernel sysfs attributes, the
+most interesting ones being:
+
+* `/sys/kernel/mm/ksm/run`: This turns KSM on (1) and off (0).
+* `/sys/kernel/mm/ksm/sleep_millisec`: This knob specifies the KSM
+scanning period.
+* `/sys/kernel/mm/ksm/pages_to_scan`: This sets the number of
+pages KSM will scan per scanning cycle.
+
+The memory density improvements that KSM can provide come at a cost
+though: depending on the number of anonymous pages it will scan,
```

Comment:
I'd drop "though". Also, should we mention the (potentially very significant) "settle time" here as an additional cost?